### PR TITLE
feat(sbb-calendar): add year and month selection views

### LINKED ISSUE #1883: story(sbb-calendar): year/month selection
### Feature Description

We want to extend the `<sbb-calendar>` component to include year/month selection.

[Figma Spec](https://www.figma.com/file/9r6xSfNmEfCFxl1yFYedrj/Lyne-Components?type=design&node-id=140%3A10&mode=design&t=OUcRyKyRxlwbPwAu-1)

**References**

- https://material.angular.io/components/datepicker/overview

**Design Spec**

- The label (month/year combination) at the top of the calendar should be an interactive element
  - Once clicked the label should switch to the available year range (with an inverted chevron) and the content should display the selectable (or disabled) years
  - Once a year has been selected, the label should show the selected year (with an inverted chevron) and the content should display the twelve months

**Technical Spec**

- If the selection of the year or month is aborted (either by pressing ESC, which should still close the datepicker, if the calendar is shown in a datepicker) or the label is clicked again, the selected date should not change
- Once a year/month has been selected, the cursor should be on the date of the month it has been before the year/month selection was started, but the date of the month is not yet selected (unless it is the same year/month as before the selection)
- If the label is clicked during the year/month selection, the year/month selection is aborted
- The year selection displays a 24 year range, where the current year or currently selected year defines the range
  - Current year => 2016 => 2039
  - Date 01.01.2011 => 1992 => 2015
  - Date 24.11.1986 => 1968 => 1991

### Definition of Done

- [ ] Component is implemented
- [ ] Documentation is provided
- [ ] Tests are implemented
- [ ] E2E Tests are implemented
- [ ] Storybook stories are implemented
- [ ] Navigation via keyboard is tested
- [ ] Screen reader output is tested
- [ ] High-contrast is tested
- [ ] Remaining accessibility is tested
- [ ] UX approved

### PR DESCRIPTION
**Pull request by DavideMininni-Fincons**
## Preflight Checklist

- [x] I have read the [Contributing Guidelines](https://github.com/lyne-design-system/lyne-components/blob/master/CONTRIBUTING.md) for this project.
- [x] I agree to follow the [Code of Conduct](https://github.com/lyne-design-system/lyne-components/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
- [x] I have searched the [pull request tracker](https://github.com/lyne-design-system/lyne-components/pulls) for a Pull Request (PR) that matches the one I want to submit, without success.

## Issue

This PR Closes #1883 

## Pull request checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)

See [Review Guidelines](../REVIEW.md) for more information what is checked during review process.

## Changes

Changes in this pull request:

- opening the sbb-calendar, it's now possible to go to the year selection view and then to the month selection view by clicking on the month label

## Browsers

I tested the build on the following browsers:

- [x] Firefox Desktop
- [x] Chrome Desktop
- [x] Edge Desktop
- [x] Safari Desktop
- [x] Chrome Mobile
- [x] Safari Mobile

## Screen readers

I tested the build on the following browsers:

- [x] JAWS Firefox Desktop
- [x] JAWS Chrome Desktop
- [x] NVDA Firefox Desktop
- [x] NVDA Chrome Desktop
- [x] VoiceOver Safari Desktop
- [x] VoiceOver Chrome Desktop
- [ ] VoiceOver Safari Mobile
- [ ] Android Accessibility Suite Chrome Mobile

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by DavideMininni-Fincons** [comment]
Notes about accessibily tests:

- JAWS with Firefox: ok
- NVDA with Chrome/Firefox and JAWS with Chrome: it's not possible to navigate with right/left arrow due to the interference with the screenreader's letter-by-letter reading
- VoiceOver with Chrome/Safari: the screenreader focus interfere with the component one; going from year selection to month, the focus is stuck in the previous position, e.g., in 2016-2039 view, if the third year is selected (2018), the focus stays in the third month (March) instead than the first. If the selected cell is beyond the 12th cell, focus stays in an undefinite position.

File: src/components/sbb-calendar/sbb-calendar.scss [hunk 1]
```diff
@@ -164,27 +189,45 @@
...
+  width: var(--sbb-calendar-cell-size);
+
+  &::before {
+    border-radius: 50%;
+  }
+}
+
+.sbb-calendar__pill {
+  width: var(--sbb-calendar-wide-cell-size);
+  border-radius: calc(var(--sbb-calendar-cell-size) / 2);
```
Comment (jeripeierSBB) [inline src/components/sbb-calendar/sbb-calendar.scss:203]: in this case we normally use "--sbb-border-radius-infinity"

File: src/components/sbb-calendar/sbb-calendar.scss [hunk 2]
```diff
@@ -164,27 +189,45 @@
...
+
+  &::before {
+    border-radius: calc(var(--sbb-calendar-cell-size) / 2);
```
Comment (jeripeierSBB) [inline src/components/sbb-calendar/sbb-calendar.scss:206]: in this case we normally use "--sbb-border-radius-infinity"

**Review by jeripeierSBB — CHANGES_REQUESTED** [review]
It's working very nice! 

I did not check every line but it looks good to me. Some open topics:
 - hover / active state of year / month button @mcilurzo 
 - when language is updated, translations of month view not updated.
 - after returning to day view, probably the focus should land on a date and not stay on month / year selection button? I just compared to material. When listening to screen reader this is the difference I noticed.
 - if in wide mode you select a month / year in the right column, the month/year will then be selected in the left column. I am not sure if this can be confusing somehow. maybe also a question for @mcilurzo .

File: src/components/sbb-calendar/sbb-calendar.tsx [hunk 3]
```diff
@@ -220,216 +297,372 @@ export class SbbCalendar implements ComponentInterface {
...
-      >
-        <thead class="sbb-calendar__table-header">
-          <tr class="sbb-calendar__table-header-row">{this._createTableHeader()}</tr>
-        </thead>
-        <tbody class="sbb-calendar__table-body">{this._createTableBody(weeks)}</tbody>
-      </table>
+  /** Creates the rows for the month selection view. */
+  private _createMonthRows(): Month[][] {
+    const shortNames: string[] = this._dateAdapter.getMonthNames('short');
+    const longNames: string[] = this._dateAdapter.getMonthNames('long');
```
[the remark this answers is not in the document]
Comment (DavideMininni-Fincons) [reply src/components/sbb-calendar/sbb-calendar.tsx]: Good catch; replaced with this._monthNames

File: src/components/sbb-calendar/sbb-calendar.tsx [hunk 4]
```diff
@@ -61,38 +102,58 @@ export class SbbCalendar implements ComponentInterface {
...
+
+  /** Grid of calendar cells representing years for the wide view. */
+  private _nextMonthYears: number[][];
+
   /** Grid of calendar cells representing the dates of the next month. */
   private _nextMonthWeeks: Day[][];
 
   /** An array containing all the month names in the current language. */
-  private _months = this._dateAdapter.getMonthNames('long');
+  private _monthNames = this._dateAdapter.getMonthNames('long');
```
[the remark this answers is not in the document]
Comment (DavideMininni-Fincons) [reply src/components/sbb-calendar/sbb-calendar.tsx]: Translation in month view should work now, the months array must be updated too in the language change event callback.

**Comment by mcilurzo** [comment]
hello @DavideMininni-Fincons, i reviewed the calendar:
https://www.figma.com/file/9r6xSfNmEfCFxl1yFYedrj?node-id=27537:271495&mode=design#548685774
https://www.figma.com/file/9r6xSfNmEfCFxl1yFYedrj?node-id=27537:271495&mode=design#548684966
https://www.figma.com/file/9r6xSfNmEfCFxl1yFYedrj?node-id=27537:271495&mode=design#548685284
thank you!

File: src/components/sbb-calendar/sbb-calendar.scss [hunk 5]
```diff
@@ -60,18 +64,40 @@
...
   &:focus-visible:not([data-focus-origin='mouse'], [data-focus-origin='touch']) {
     @include sbb.focus-outline;
 
     outline-offset: var(--sbb-spacing-fixed-1x);
-    border-radius: var(--sbb-border-radius-2x);
+  }
+
+  @include sbb.hover-mq {
+    &:not(:active, :disabled):hover {
+      padding-block-end: 0.0625rem;
```
Comment (jeripeierSBB) [inline src/components/sbb-calendar/sbb-calendar.scss:95]: I remember we once had flickering problems when shifting content with padding when using screens with worse resolution than our mac books :). I can't test it now, but maybe we should test that too. The way to go was to use transform translation then.

Comment (MarioCastigliano) [reply src/components/sbb-calendar/sbb-calendar.scss:95]: Should I apply the same change also on the day cell?

Comment (jeripeierSBB) [reply src/components/sbb-calendar/sbb-calendar.scss:95]: if you would change one, then also the other please ;) can you additionally use pxtorem function?
But as written above, I did not test your scenario, maybe it's not even a problem, I was just remembering the other case (which was button and header action as I remember)

**Review by kyubisation — COMMENTED** [review]
Generally looks good to me, nice work! 😃 
I feel like the focus/pre-selection is not yet working as I would expect? (Tested with https://lyne-components-preview-pr1951.app.sbb.ch/?path=/story/components-sbb-datepicker-sbb-datepicker--in-form-field )
Selecting a year switches to the months view, but the focus seems now to be on January and is also not highlighted/black.

**Review by kyubisation — APPROVED** [review]
LGTM 👍 
Thank you for your work! 😃 
@MarioCastigliano, @DavideMininni-Fincons Quick question for long-term; Should we keep this implementation as it is or should we in the future evaluate splitting this into sub-components (e.g. similar to the Angular Material implementation)?

**Comment by MarioCastigliano** [comment]
Thanks @kyubisation for the review, after discussing it with @DavideMininni-Fincons, we agreed that given the upcoming migration, it's best to leave it as is and reconsider the matter after the migration is complete

**Pull request merged**